File: backend/indeed_scraper.py
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from webdriver_manager.chrome import ChromeDriverManager
import time
import random
import re

EXPERIENCE_FILTERS = {
    "0-1":  "entry_level",
    "1-3":  "entry_level",
    "3-5":  "mid_level",
    "5-10": "senior_level",
    "10+":  "senior_level",
    "internship": "internship",
    "entry":  "entry_level",
    "mid":    "mid_level",
    "senior": "senior_level",
}

COUNTRY_DOMAINS = {
    "india": "in.indeed.com",
    "uk": "uk.indeed.com",
    "united kingdom": "uk.indeed.com",
    "canada": "ca.indeed.com",
    "australia": "au.indeed.com",
    "germany": "de.indeed.com",
    "france": "fr.indeed.com",
    "singapore": "sg.indeed.com",
    "uae": "ae.indeed.com",
    "united arab emirates": "ae.indeed.com",
}

# Maps user experience range → (min_years, max_years) for post-filter
EXP_RANGE = {
    "0-1":  (0, 1),
    "1-3":  (1, 3),
    "3-5":  (3, 5),
    "5-10": (5, 10),
    "10+":  (10, 99),
}


def collect_indeed_jobs(keywords, experience_level="0-1", location=""):
    jobs = []
    exp_filter = EXPERIENCE_FILTERS.get(experience_level, "entry_level")
    exp_range  = EXP_RANGE.get(experience_level)

    clean_keywords = _clean_keywords(keywords)
    print(f"[Indeed] Keywords: {clean_keywords} | Exp: {experience_level} | Location: {location}")

    base_domain = "www.indeed.com"
    loc_param = ""
    if location:
        loc_lower = location.lower()
        for country_key, domain in COUNTRY_DOMAINS.items():
            if country_key in loc_lower:
                base_domain = domain
                break
        loc_param = location.replace(" ", "+")

    chrome_options = Options()
    chrome_options.add_argument("--headless=new")
    chrome_options.add_argument("--no-sandbox")
    chrome_options.add_argument("--disable-dev-shm-usage")
    chrome_options.add_argument("--disable-gpu")
    chrome_options.add_argument("--window-size=1920,1080")
    chrome_options.add_argument("--disable-blink-features=AutomationControlled")
    chrome_options.add_experimental_option("excludeSwitches", ["enable-automation"])
    chrome_options.add_experimental_option("useAutomationExtension", False)
    chrome_options.add_argument(
        "user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
    )

    driver = webdriver.Chrome(
        service=Service(ChromeDriverManager().install()),
        options=chrome_options
    )
    driver.execute_script(
        "Object.defineProperty(navigator, 'webdriver', {get: () => undefined})"
    )

    try:
        for keyword in clean_keywords:
            print(f"[Indeed] Searching: '{keyword}' on {base_domain}")

            search_url = (
                f"https://{base_domain}/jobs"
                f"?q={keyword.replace(' ', '+')}"
                f"&explvl={exp_filter}&sort=date"
            )
            if loc_param:
                search_url += f"&l={loc_param}"

            driver.get(search_url)
            time.sleep(random.uniform(4, 6))

            title_check = driver.title
            print(f"  Page: {title_check}")

            if "Blocked" in title_check or "Just a moment" in title_check:
                time.sleep(10)
                driver.get(search_url)
                time.sleep(5)
                if "Blocked" in driver.title or "Just a moment" in driver.title:
                    print(f"  ❌ Blocked — skipping")
                    continue

            cards = driver.find_elements(By.CSS_SELECTOR, ".job_seen_beacon")
            if not cards:
                cards = driver.find_elements(By.CSS_SELECTOR, "[data-jk]")

            print(f"  Cards: {len(cards)}")
            if not cards:
                continue

            for card in cards[:15]:  # scrape more, filter down after
                try:
                    # Title
                    title = ""
                    for sel in ["h2.jobTitle span[title]", "h2.jobTitle span", "a.jcs-JobTitle span"]:
                        try:
                            el = card.find_element(By.CSS_SELECTOR, sel)
                            title = el.get_attribute("title") or el.text.strip()
                            if title:
                                break
                        except:
                            continue

                    # Company
                    company = ""
                    for sel in ["[data-testid='company-name']", ".companyName"]:
                        try:
                            company = card.find_element(By.CSS_SELECTOR, sel).text.strip()
                            if company:
                                break
                        except:
                            continue

                    # Location
                    loc = ""
                    for sel in ["[data-testid='text-location']", ".companyLocation"]:
                        try:
                            loc = card.find_element(By.CSS_SELECTOR, sel).text.strip()
                            if loc:
                                break
                        except:
                            continue

                    # Salary
                    salary = ""
                    for sel in [
                        "[data-testid='attribute_snippet_testid']",
                        ".salary-snippet-container",
                        ".salaryOnly",
                        "[class*='SalarySnippet']",
                        "[class*='salary']",
                    ]:
                        try:
                            el = card.find_element(By.CSS_SELECTOR, sel)
                            text = el.text.strip()
                            if text and any(c in text for c in ["$", "₹", "£", "€", "year", "hour", "month", "per"]):
                                salary = text
                                break
                        except:
                            continue

                    if not salary:
                        try:
                            for meta in card.find_elements(By.CSS_SELECTOR, ".metadata, li"):
                                t = meta.text.strip()
                                if t and any(c in t for c in ["$", "₹", "£", "LPA", "lakh", "k per"]):
                                    salary = t
                                    break
                        except:
                            pass

                    # Snippet text — used for experience filtering
                    snippet = ""
                    try:
                        snippet = card.text.lower()
                    except:
                        pass

                    # Apply link
                    link = ""
                    try:
                        a_el = card.find_element(By.CSS_SELECTOR, "a.jcs-JobTitle, h2.jobTitle a")
                        jk = a_el.get_attribute("data-jk")
                        if jk:
                            link = f"https://{base_domain}/viewjob?jk={jk}"
                        else:
                            href = a_el.get_attribute("href") or ""
                            link = href.split("?")[0] if href else ""
                    except:
                        pass

                    easy_apply = False
                    try:
                        easy_apply = "easily apply" in snippet
                    except:
                        pass

                    if not title or not link:
                        continue

                    # ── Post-scrape experience filter ──────────────────────
                    # Extract years mentioned in the card snippet
                    # e.g. "5+ years", "3-5 years", "minimum 2 years"
                    if exp_range:
                        mentioned = _extract_years_from_text(snippet)
                        if mentioned is not None:
                            min_req, max_req = mentioned
                            user_min, user_max = exp_range
                            # Reject if job requires MORE experience than user's max
                            # or if job requires MUCH LESS than user's min (more than 2yr gap)
                            if min_req > user_max:
                                print(f"  ⛔ Filtered out: '{title}' requires {min_req}+ yrs, user wants {user_min}-{user_max}")
                                continue
                            if max_req is not None and max_req < user_min - 1:
                                print(f"  ⛔ Filtered out: '{title}' requires ≤{max_req} yrs, user wants {user_min}-{user_max}")
                                continue

                    jobs.append({
                        "title": title,
                        "company": company,
                        "location": loc,
                        "salary": salary,
                        "experience_level": experience_level,
                        "apply_link": link,
                        "easy_apply": easy_apply,
                        "source": "Indeed",
                    })
                    print(f"  ✅ {title} @ {company}")

                except Exception as e:
                    print(f"  Card error: {e}")
                    continue

            time.sleep(random.uniform(2, 4))

    finally:
        driver.quit()

    print(f"[Indeed] Total after filtering: {len(jobs)} jobs")
    return jobs


def _extract_years_from_text(text):
    """
    Extract experience requirement from job card text.
    Returns (min_years, max_years) or None if not mentioned.

    Handles patterns like:
    - "5+ years"
    - "3-5 years"
    - "minimum 2 years"
    - "at least 3 years"
    - "1 to 3 years"
    - "5 years of experience"
    """
    text = text.lower()

    # Pattern: "X+ years" or "X or more years"
    m = re.search(r'(\d+)\s*\+\s*years?', text)
    if m:
        n = int(m.group(1))
        return (n, 99)

    # Pattern: "X-Y years" or "X to Y years"
    m = re.search(r'(\d+)\s*(?:-|to)\s*(\d+)\s*years?', text)
    if m:
        return (int(m.group(1)), int(m.group(2)))

    # Pattern: "minimum/at least/over X years"
    m = re.search(r'(?:minimum|at least|over|more than)\s+(\d+)\s*years?', text)
    if m:
        n = int(m.group(1))
        return (n, 99)

    # Pattern: "X years of experience" (single number)
    m = re.search(r'(\d+)\s+years?\s+(?:of\s+)?(?:experience|exp)', text)
    if m:
        n = int(m.group(1))
        return (n, n)

    return None  # no experience mentioned — don't filter


def _clean_keywords(keywords):
    VALID = [
        "analyst", "engineer", "developer", "scientist", "manager",
        "designer", "architect", "consultant", "specialist", "intern",
        "data", "software", "machine learning", "frontend", "backend",
        "fullstack", "devops", "cloud", "python", "java", "power bi",
        "business", "product", "project", "network", "security", "ai", "ml"
    ]
    cleaned, seen = [], set()
    for kw in keywords:
        kw = kw.replace("\n", " ").strip()
        kw_lower = kw.lower()
        if kw_lower in seen:
            continue
        if len(kw.split()) == 1 and kw_lower in {
            "time", "hands", "enterprise", "team", "good",
            "strong", "experience", "knowledge", "skills"
        }:
            continue
        if not any(v in kw_lower for v in VALID):
            continue
        cleaned.append(kw)
        seen.add(kw_lower)
    return cleaned if cleaned else ["data analyst"]

File: backend/src/scrapers/indeed.scraper.js
/**
 * scrapers/indeed.scraper.js
 * ──────────────────────────────────────────────────────────────────────────
 * Fetches Indeed job listings via their public search — no puppeteer.
 * Uses the RSS/JSON endpoint to avoid Render IP blocks.
 * ──────────────────────────────────────────────────────────────────────────
 */

"use strict";

const https = require("https");
const http  = require("http");
const {
  extractYearsFromText,
  formatExpRequired,
  checkExpMismatch,
  getUserExpRange,
} = require("../utils/expParser");

const EXPERIENCE_FILTERS = {
  "0-1":  "entry_level",
  "1-3":  "entry_level",
  "3-5":  "mid_level",
  "5-10": "senior_level",
  "10+":  "senior_level",
};

const COUNTRY_DOMAINS = {
  india:                  "in.indeed.com",
  uk:                     "uk.indeed.com",
  "united kingdom":       "uk.indeed.com",
  canada:                 "ca.indeed.com",
  australia:              "au.indeed.com",
  germany:                "de.indeed.com",
  singapore:              "sg.indeed.com",
  uae:                    "ae.indeed.com",
  "united arab emirates": "ae.indeed.com",
};

// ── HTTP GET with redirect following ─────────────────────────────────────────
function fetchUrl(url, headers = {}, redirects = 3) {
  return new Promise((resolve, reject) => {
    const lib = url.startsWith("https") ? https : http;
    const req = lib.get(url, {
      headers: {
        "User-Agent":
          "Mozilla/5.0 (Windows NT 10.0; Win64; x64) " +
          "AppleWebKit/537.36 (KHTML, like Gecko) " +
          "Chrome/124.0.0.0 Safari/537.36",
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        "Accept-Language": "en-US,en;q=0.9",
        "Accept-Encoding": "gzip, deflate",
        ...headers,
      },
      timeout: 20000,
    }, (res) => {
      // Follow redirects
      if ([301, 302, 303, 307, 308].includes(res.statusCode) && res.headers.location && redirects > 0) {
        const next = res.headers.location.startsWith("http")
          ? res.headers.location
          : new URL(res.headers.location, url).href;
        resolve(fetchUrl(next, headers, redirects - 1));
        return;
      }

      // Handle gzip
      let stream = res;
      if (res.headers["content-encoding"] === "gzip") {
        const zlib = require("zlib");
        stream = res.pipe(zlib.createGunzip());
      }

      let data = "";
      stream.on("data", (c) => (data += c));
      stream.on("end", () => resolve({ status: res.statusCode, body: data }));
      stream.on("error", reject);
    });
    req.on("error", reject);
    req.on("timeout", () => { req.destroy(); reject(new Error("Request timeout")); });
  });
}

function stripHtml(str = "") {
  return str.replace(/<[^>]*>/g, " ").replace(/&amp;/g, "&").replace(/&nbsp;/g, " ").replace(/\s+/g, " ").trim();
}

// ── Main export ───────────────────────────────────────────────────────────────
async function collectIndeedJobs(keywords, experienceLevel = "0-1", location = "") {
  const { min: userMin, max: userMax } = getUserExpRange(experienceLevel);
  const expFilter = EXPERIENCE_FILTERS[experienceLevel] || "entry_level";
  const cleanKws  = cleanKeywords_(keywords);

  let baseDomain = "www.indeed.com";
  const locLower = location.toLowerCase();
  for (const [key, domain] of Object.entries(COUNTRY_DOMAINS)) {
    if (locLower.includes(key)) { baseDomain = domain; break; }
  }
  const locParam = location.replace(/\s+/g, "+");

  console.log(`[Indeed] Keywords: ${cleanKws} | Exp: ${experienceLevel} | Location: ${location}`);

  const jobs = [];

  for (const keyword of cleanKws) {
    try {
      // Use Indeed's RSS feed — much less likely to be blocked than the main page
      const rssUrl =
        `https://${baseDomain}/rss` +
        `?q=${encodeURIComponent(keyword)}` +
        `&explvl=${expFilter}` +
        `&sort=date` +
        (locParam ? `&l=${locParam}` : "");

      console.log(`[Indeed] RSS: ${rssUrl}`);

      const { status, body } = await fetchUrl(rssUrl, {
        "Referer": `https://${baseDomain}/`,
      });

      if (status !== 200 || !body.includes("<item>")) {
        console.log(`[indeed] RSS returned ${status} or no items — trying HTML fallback`);

        // HTML fallback
        const htmlUrl =
          `https://${baseDomain}/jobs` +
          `?q=${encodeURIComponent(keyword)}` +
          `&explvl=${expFilter}&sort=date` +
          (locParam ? `&l=${locParam}` : "");

        const fallback = await fetchUrl(htmlUrl).catch(() => ({ status: 0, body: "" }));
        if (!fallback.body || fallback.status !== 200) {
          console.log(`[indeed] ✗ HTML fallback also failed`);
          continue;
        }

        // Parse HTML job cards
        const htmlJobs = parseIndeedHTML(fallback.body, baseDomain, userMin, userMax, experienceLevel);
        jobs.push(...htmlJobs);
        console.log(`[Indeed] HTML fallback: ${htmlJobs.length} jobs`);
        continue;
      }

      // Parse RSS
      const itemRegex = /<item>([\s\S]*?)<\/item>/g;
      let match;
      const found = [];

      while ((match = itemRegex.exec(body)) !== null && found.length < 10) {
        const item = match[1];

        const title   = stripHtml(item.match(/<title>([\s\S]*?)<\/title>/)?.[1] || "");
        const company = stripHtml(item.match(/<source[^>]*>([\s\S]*?)<\/source>/)?.[1] || "");
        const link    = (item.match(/<link>([\s\S]*?)<\/link>/) || item.match(/<guid[^>]*>([\s\S]*?)<\/guid>/))?.[1]?.trim() || "";
        const desc    = stripHtml(item.match(/<description>([\s\S]*?)<\/description>/)?.[1] || "").toLowerCase();
        const locRaw  = stripHtml(item.match(/<[^>]*location[^>]*>([\s\S]*?)<\/[^>]+>/i)?.[1] || location);

        if (!title || !link) continue;

        const expRange    = extractYearsFromText(desc);
        const expReq      = formatExpRequired(expRange);
        const expVerified = expRange !== null;
        const { hardDrop, mismatch, hardMismatch } = checkExpMismatch(expRange, userMin, userMax);

        if (hardDrop) {
          console.log(`  [P1] Hard-drop '${title}': ${expReq}`);
          continue;
        }

        console.log(`  ${expVerified ? "✅[P1]" : "🔍[P2]"} ${title} @ ${company} | req: ${expReq || "unknown"}`);

        found.push({
          title,
          company,
          location:          locRaw || location,
          salary:            "",
          experience_level:  experienceLevel,
          exp_required:      expReq,
          exp_mismatch:      mismatch,
          exp_hard_mismatch: hardMismatch,
          exp_verified:      expVerified,
          apply_link:        link,
          easy_apply:        false,
          source:            "Indeed",
        });
      }

      console.log(`[Indeed] ${found.length} jobs from RSS`);
      jobs.push(...found);

    } catch (err) {
      console.log(`[indeed] ✗ Error: ${err.message}`);
    }

    await new Promise((r) => setTimeout(r, 1000 + Math.random() * 1000));
  }

  console.log(`[Indeed] Total: ${jobs.length} jobs`);
  return jobs;
}

// ── HTML parser fallback ──────────────────────────────────────────────────────
function parseIndeedHTML(html, domain, userMin, userMax, experienceLevel) {
  const found = [];
  const cardRegex = /data-jk="([^"]+)"[\s\S]*?<h2[^>]*jobTitle[^>]*>([\s\S]*?)<\/h2>[\s\S]*?class="[^"]*companyName[^"]*"[^>]*>([\s\S]*?)<\/(?:span|a)>/g;
  let match;
  while ((match = cardRegex.exec(html)) !== null && found.length < 10) {
    const jk      = match[1];
    const title   = stripHtml(match[2]);
    const company = stripHtml(match[3]);
    const link    = `https://${domain}/viewjob?jk=${jk}`;

    if (!title) continue;

    found.push({
      title,
      company,
      location:          "",
      salary:            "",
      experience_level:  experienceLevel,
      exp_required:      null,
      exp_mismatch:      false,
      exp_hard_mismatch: false,
      exp_verified:      false,
      apply_link:        link,
      easy_apply:        false,
      source:            "Indeed",
    });
  }
  return found;
}

// ── Keyword cleaner ───────────────────────────────────────────────────────────
function cleanKeywords_(keywords) {
  const VALID = [
    "analyst","engineer","developer","scientist","manager","designer",
    "architect","consultant","specialist","intern","data","software",
    "machine learning","frontend","backend","fullstack","devops","cloud",
    "python","java","power bi","business","product","project","network",
    "security","ai","ml",
  ];
  const NOISE = new Set(["time","hands","enterprise","team","good","strong","experience","knowledge","skills"]);
  const seen  = new Set();
  const out   = [];
  for (const kw of keywords) {
    const clean = kw.replace(/\n/g, " ").trim();
    const lower = clean.toLowerCase();
    if (seen.has(lower)) continue;
    if (clean.split(" ").length === 1 && NOISE.has(lower)) continue;
    if (!VALID.some((v) => lower.includes(v))) continue;
    seen.add(lower);
    out.push(clean);
  }
  return out.length ? out : ["data analyst"];
}

module.exports = { collectIndeedJobs };
